selene-lib/src/rules/mismatched_arg_count.rs

```diff
@@ -0,0 +1,425 @@
+    /// pass the lint, since the `nil` could be implicitly provided.
+    pub fn correct_num_args_provided(&self, provided: PassedArgumentCount) -> bool {
+        match self {
+            ParameterCount::Fixed(required) => match provided {
+                PassedArgumentCount::Fixed(provided) => provided <= *required,
+                // If we have function(a, b, c), but we provide foo(a, call()), we cannot infer anything
+                // but if we provide foo(a, b, c, call()), we know we have too many
+                PassedArgumentCount::Variable(atleast_provided) => atleast_provided <= *required,
+            },
+            // function(a, b, ...) - if we call it through foo(a), b and the varargs could be implicitly nil.
+            // there is no upper bound since foo(a, b, c, d) is valid - therefore any amount of arguments provided is valid
+            ParameterCount::Minimum(_) => true,
+            // Any amount of arguments could be provided
+            ParameterCount::Variable => true,
+        }
+    }
+
+    pub fn to_message(&self, provided: PassedArgumentCount) -> String {
+        match self {
+            ParameterCount::Fixed(required) => {
+                format!("{} arguments but {} were supplied", required, provided)
+            }
+            ParameterCount::Minimum(required) => format!(
+                "atleast {} arguments but {} were supplied",
+                required, provided
+            ),
+            ParameterCount::Variable => "a variable amount of arguments".to_owned(),
+        }
+    }
+}
+impl Display for ParameterCount {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ParameterCount::Fixed(required) => write!(f, "expected {} arguments", required),
+            ParameterCount::Minimum(required) => {
+                write!(f, "expected atleast {} arguments", required)
+            }
+            ParameterCount::Variable => write!(f, "expected any number of arguments"),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+enum PassedArgumentCount {
+    /// Passed a fixed amount of arguments, such as foo(a, b, c) or foo(a, call(), c) or foo(a, ..., c)
+    Fixed(usize),
+    /// Passed a variable of arguments - but we know the lower bound: e.g. foo(a, b, call()) or foo(a, b, ...)
+    Variable(usize),
+}
+impl Display for PassedArgumentCount {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            PassedArgumentCount::Fixed(amount) => write!(f, "{} arguments", amount),
+            PassedArgumentCount::Variable(amount) => write!(f, "atleast {} arguments", amount),
+        }
+    }
+}
+
+fn function_call_argument_count(function_args: &ast::FunctionArgs) -> PassedArgumentCount {
+    match function_args {
+        ast::FunctionArgs::Parentheses { arguments, .. } => {
+            // We need to be wary of items with side effects, such as function calls or ... being the last argument passed
```

Comment:
`...` isn't marked as having side-effects, and I don't see this in your tests. I suspect this doesn't work the way you think it does.

Reply:
Technically this actually doesn't really matter anymore since we are only linting for too-many-arguments, rather than too-little-arguments, so this isn't particularly important anymore.
However, I will fix this for completeness